### PR TITLE
chore(base-driver): Update HTTP caching logic and align it with the standard

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
## Proposed changes

Now HTTP caching logic properly follows https://developer.mozilla.org/en-US/docs/Web/HTTP/Caching, which allows to reduce the amount of requests and the boilerplate code.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

### CODE REVIEW

File: packages/base-driver/lib/basedriver/helpers.js [hunk 1]
```diff
@@ -2,16 +2,21 @@ import _ from 'lodash';
...
+import B from 'bluebird';
 
 const IPA_EXT = '.ipa';
 const ZIP_EXTS = ['.zip', IPA_EXT];
 const ZIP_MIME_TYPES = ['application/zip', 'application/x-zip-compressed', 'multipart/x-zip'];
 const CACHED_APPS_MAX_AGE = 1000 * 60 * 60 * 24; // ms
 const MAX_CACHED_APPS = 1024;
+const HTTP_STATUS_NOT_MODIFIED = 304;
+const DEFAULT_REQ_HEADERS = Object.freeze({
+  'user-agent': 'Appium',
```
Comment (boneskull) [inline packages/base-driver/lib/basedriver/helpers.js:18]: guess it doesn't really matter but might want to stuff a version in there too.

File: packages/base-driver/lib/basedriver/helpers.js [hunk 2]
```diff
@@ -250,90 +185,113 @@ async function configureApp(app, options = /** @type {ConfigureAppOptions} */ ({
...
-          logger.info(
-            `The current file extension '${resultingExt}' is not supported. ` +
+        if (headers['content-disposition'] && /^attachment/i.test(headers['content-disposition'])) {
+          logger.debug(`Content-Disposition: ${headers['content-disposition']}`);
+          const match = /filename="([^"]+)/i.exec(headers['content-disposition']);
+          if (match) {
+            fileName = fs.sanitizeName(match[1], {
+              replacement: SANITIZE_REPLACEMENT,
+            });
+            shouldUnzipApp = shouldUnzipApp || ZIP_EXTS.includes(path.extname(fileName));
```
Comment (boneskull) [inline packages/base-driver/lib/basedriver/helpers.js:267]: `ZIP_EXTS` should be a `Set`

File: packages/base-driver/lib/basedriver/helpers.js [hunk 3]
```diff
@@ -429,16 +388,76 @@ async function configureApp(app, options = /** @type {ConfigureAppOptions} */ ({
...
+    throw new Error(`Cannot fetch the application: ${err.message}`);
+  }
+
+  const secondsElapsed = timer.getDuration().asSeconds;
+  const {size} = await fs.stat(dstPath);
+  logger.debug(
+    `The application (${util.toReadableSizeString(size)}) ` +
+    `has been downloaded to '${dstPath}' in ${secondsElapsed.toFixed(3)}s`
+  );
+  if (secondsElapsed >= 2) {
```
Comment (boneskull) [inline packages/base-driver/lib/basedriver/helpers.js:455]: magic number

File: packages/base-driver/lib/basedriver/helpers.js [hunk 4]
```diff
@@ -429,16 +388,76 @@ async function configureApp(app, options = /** @type {ConfigureAppOptions} */ ({
...
+ * @returns {Promise<string>} The same dstPath
+ * @throws {Error} If there was a failure while downloading the file
+ */
+async function fetchApp (srcStream, dstPath) {
+  const timer = new timing.Timer().start();
+  try {
+    const writer = fs.createWriteStream(dstPath);
+    srcStream.pipe(writer);
+
+    await new B((resolve, reject) => {
```
Comment (boneskull) [inline packages/base-driver/lib/basedriver/helpers.js:437]: consider using `pipeline` here from `node:stream/promises` after dropping node v14

File: packages/base-driver/lib/basedriver/helpers.js [hunk 5]
```diff
@@ -613,4 +632,12 @@ export {configureApp, isPackageOrBundle, duplicateKeys, parseCapsArray, generate
  * @property {Date?} lastModified
  * @property {boolean} immutable
  * @property {number?} maxAge
+ * @property {string?} etag
+ */
+
+/**
+ * @typedef RemoteAppData
```
Comment (boneskull) [inline packages/base-driver/lib/basedriver/helpers.js:639]: please add a description and descriptions of props

**Review by boneskull — APPROVED** [review]
I can't really comment on the intent of this or if it's "correct" to make these changes, so just had some suggestions.

File: packages/base-driver/lib/basedriver/helpers.js (hunk 3 above)
Comment (mykola-mokhnach) [reply packages/base-driver/lib/basedriver/helpers.js:455]: it does not make sense for me to measure approximate speed if the download itself took less than two seconds. Would it be better to just add a comment there or you could propose a good constant name for the minimum number of seconds, which is `2`?

File: packages/base-driver/lib/basedriver/helpers.js (hunk 4 above)
Comment (mykola-mokhnach) [reply packages/base-driver/lib/basedriver/helpers.js:437]: I did not know about such thing

File: packages/base-driver/lib/basedriver/helpers.js (hunk 1 above)
Comment (mykola-mokhnach) [reply packages/base-driver/lib/basedriver/helpers.js:18]: do we have some helper, which allows to fetch the version number?

File: packages/base-driver/lib/basedriver/helpers.js (hunk 2 above)
Comment (mykola-mokhnach) [reply packages/base-driver/lib/basedriver/helpers.js:267]: changed

File: packages/base-driver/lib/basedriver/helpers.js (hunk 5 above)
Comment (mykola-mokhnach) [reply packages/base-driver/lib/basedriver/helpers.js:639]: np, added descriptions

File: packages/base-driver/lib/basedriver/helpers.js (hunk 1 above)
Comment (jlipps) [reply packages/base-driver/lib/basedriver/helpers.js:18]: could just use basedriver's version number so you don't have to look outside this package

```
Appium (BaseDriver vX.Y.Z)
```

File: packages/base-driver/lib/basedriver/helpers.js (hunk 3 above)
Comment (jlipps) [reply packages/base-driver/lib/basedriver/helpers.js:456]: `DOWNLOAD_MEASUREMENT_THRESHOLD`?

File: packages/base-driver/lib/basedriver/helpers.js (hunk 1 above)
Comment (mykola-mokhnach) [reply packages/base-driver/lib/basedriver/helpers.js:18]: probably an example of getting the version of base driver package?

Comment (jlipps) [reply packages/base-driver/lib/basedriver/helpers.js:18]: something like `require('../path/to/package.json').version`

Comment (mykola-mokhnach) [reply packages/base-driver/lib/basedriver/helpers.js:18]: added

File: packages/base-driver/lib/basedriver/helpers.js (hunk 3 above)
Comment (mykola-mokhnach) [reply packages/base-driver/lib/basedriver/helpers.js:456]: changed

**Pull request merged**